[Spring 지하철 노선도 - 1,2단계] 리버(구동희) 미션 제출합니다.

gudonghee2000 opened the pull request:
안녕하세요 던!
이번 리뷰를 부탁드릴 리버입니다!
저번 로또미션때 뵙고 다시한번 뵙게되었어요!

아직 스프링을 잘 다루지 못하지만
이번 리뷰 잘부탁드립니다 던!
감사합니다😊

jaejae-yoo committed:
docs: 기능 요구사항 작성

jaejae-yoo committed:
feat: 지하철 역 관리 API 기능

jaejae-yoo committed:
feat: 지하철 노선 등록 기능

jaejae-yoo committed:
feat: 지하철 노선 목록 기능

jaejae-yoo committed:
feat: 지하철 노선 조회

jaejae-yoo committed:
feat: 지하철 노선 수정

jaejae-yoo committed:
feat: 지하철 노선 삭제

jaejae-yoo committed:
feat: Station 관리 Spring 애노테이션 적용

jaejae-yoo committed:
feat: Line 관리 Spring 애노테이션 적용

jaejae-yoo committed:
refactor: 불필요한 라이브러리 제거 및 공백 수정

jaejae-yoo committed:
docs:기능 요구사항 업데이트

gch01410 commented on src/test/java/wooteco/subway/acceptance/LineAcceptanceTest.java:
```diff
@@ -0,0 +1,187 @@
+        assertThat(response.header("Location")).isNotBlank();
+    }
+
+    @DisplayName("지하철 노선을 조회한다.")
+    @Test
+    void getLines() {
```
테스트 코드가 실패하고 있어요! 확인 부탁드릴게요~

gudonghee2000 replied:
@Sql 애노테이션을 통해 실패하는 테스트 코드를 수정하였어요!

gch01410 commented on src/test/java/wooteco/subway/acceptance/StationAcceptanceTest.java:
```diff
@@ -1,25 +1,27 @@
 @DisplayName("지하철역 관련 기능")
 public class StationAcceptanceTest extends AcceptanceTest {
+
     @DisplayName("지하철역을 생성한다.")
     @Test
```
여기도 테스트가 실패하고 있어요! 확인 부탁드릴게요!

gudonghee2000 replied:
@Sql 애노테이션을 통해 실패하는 테스트 코드를 수정하였어요!

gch01410 commented on src/main/java/wooteco/subway/ui/StationController.java:
```diff
@@ -1,39 +1,40 @@
+        this.stationService = stationService;
+    }
+
     @PostMapping("/stations")
```
`@RequestMapping`으로 경로중복을 제거해볼까요?

gudonghee2000 replied:
경로중복을 제거하였어요!

gch01410 commented on src/main/java/wooteco/subway/ui/StationController.java:
```diff
@@ -1,39 +1,40 @@
+        return ResponseEntity.created(URI.create("/stations/" + stationResponse.getId())).body(stationResponse);
     }
 
     @GetMapping(value = "/stations", produces = MediaType.APPLICATION_JSON_VALUE)
     public ResponseEntity<List<StationResponse>> showStations() {
```
이 API에만 produces 값을 설정한 이유가 있을까요~?

gudonghee2000 replied:
해당 부분에서 produces 값이 설정되어있는 것은 처음 fork시 만들어진 파일에서 설정된 부분이예요 던!
아마, 데이터를 전달 받는 view 부분에서 json으로 된 데이터만 전달 받도록 하기 위해서 설정되어있는 것은 아닐까 생각해요!

gch01410 commented on src/main/java/wooteco/subway/service/LineService.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    private void validateDuplicateLine(LineRequest lineRequest) {
+        Optional<Line> optional = lineDao.findAll()
+                .stream()
+                .filter(line -> line.getName().equals(lineRequest.getName()))
+                .findAny();
```
스트림의 `anyMatch()`를 사용할 수 있어보여요!

gudonghee2000 replied:
`anyMatch()`를 사용 하도록 변경해보았어요!

gch01410 commented on src/main/java/wooteco/subway/exception/ClientException.java:
```diff
@@ -0,0 +1,8 @@
+package wooteco.subway.exception;
+
+public class ClientException extends RuntimeException {
+
+    public ClientException(String message) {
+        super(message);
+    }
```
커스텀 예외를 사용하셨군요! 👍

gch01410 commented on src/main/java/wooteco/subway/ui/StationControllerAdvice.java:
```diff
@@ -0,0 +1,16 @@
+@RestControllerAdvice
+public class StationControllerAdvice {
+
+    @ExceptionHandler({ClientException.class})
+    public ResponseEntity<String> IllegalArgumentExceptionHandle(ClientException exception) {
+        return ResponseEntity.badRequest().body(exception.getMessage());
+    }
```
ClientException 외에 다른 예외가 발생하는 경우의 처리도 있어야하지 않을까요?

gudonghee2000 replied:
추가적으로 StationService와 LineService에서 발생 할 수 있는 예외  처리를 추가하였어요!

gch01410 commented on src/test/java/wooteco/subway/dao/LineDaoImplTest.java:
```diff
@@ -0,0 +1,86 @@
+    void setUp() {
+        lineDao = new LineDaoImpl(jdbcTemplate);
+
+        jdbcTemplate.execute("DROP TABLE line IF EXISTS");
+        jdbcTemplate.execute("create table if not exists LINE(\n" +
+                "id bigint auto_increment not null,\n" +
+                "name varchar(255) not null unique,\n" +
+                "color varchar(20) not null,\n" +
+                "primary key(id));");
```
테스트 코드에서도 schema.sql 파일을 추가해서 테이블을 자동으로 생성할 수 있을 것 같아요!
아래 사이트를 참고해봐요!
https://cheese10yun.github.io/sql-test/

gudonghee2000 replied:
보내주신 자료를 통해서 @Sql 애노테이션을 활용하여 test 스키마를 자동 생성하도록 변경해보았어요 던!

gch01410 commented on src/test/java/wooteco/subway/acceptance/LineAcceptanceTest.java:
```diff
@@ -0,0 +1,187 @@
+
+        // then
+        assertThat(newResponse.statusCode()).isEqualTo(HttpStatus.OK.value());
+        assertThat(resultLineId).isEqualTo(newResponse.jsonPath().getLong("id"));
```
reponse body의 값도 확인해볼 수 있을까요?

gudonghee2000 replied:
response body도 테스트 하도록 변경해보았어요!

gch01410 reviewed (changes requested):
안녕하세요 리버! 다시 만났네요 ㅎㅎ
전반적으로 깔끔하게 구현해주셨어요! 💯 
코멘트 남겼으니 확인 부탁드려요!

gudonghee2000 committed:
refactor: StationController의 중복되는 API Mapping을 RequestMapping으로 변경

gudonghee2000 committed:
refactor: LineController의 중복되는 API Mapping을 RequestMapping으로 변경

gudonghee2000 committed:
refactor: 지하철역과 지하철 노선이 중복되는 경우를 잡는 로직 변경

gudonghee2000 committed:
test: test 패키지의 schema.sql을 사용하여 Dao 객체 테스트

gudonghee2000 committed:
test: Service 객체 테스트 조건 추가

gudonghee2000 committed:
test: E2E테스트와 Dao 테스트에 @Sql을 통해 test Schema 연결

gudonghee2000 committed:
test: E2E테스트에서 response body도 테스트하도록 변경

gudonghee2000 committed:
test: Dao 테스트의 부모객체 생성

gudonghee2000 committed:
refactor: 예외발생시 커스텀예외 추가

gudonghee2000 committed:
feat: 지하철역 삭제요청시 지하철역 정보가 DB에 없으면 예외발생

gudonghee2000 committed:
feat: LineService 예외 발생 추가

1. id를 통해 지하철 노선 조회시 id가 DB에 존재하지 않으면 예외발생
2. id를 통해 지하철 노선 삭제시 id가 DB에 존재하지 않으면 예외발생
3. id를 통해 지하철 노선 시 업데이트시 id가 DB에 존재하지 않으면 예외발생
4. 지하철 노선 업데이트시 중복되는 이름으로 업데이트시 예외발생

gch01410 commented on src/main/java/wooteco/subway/service/StationService.java:
```diff
@@ -0,0 +1,60 @@
+        try {
+            stationDao.findById(id);
+        } catch (DataNotFoundException e) {
+            throw new DataNotFoundException("존재하지 않는 역입니다.");
```
커스텀 예외 👍 
엔티티별로 만들어 줄 수도 있어요!
`StationNotFoundException`, `LineNotFoundException`

gch01410 reviewed (approved):
안녕하세요 리버!
실패하는 테스트 코드가 있는데 확인부탁드릴게요.
다음 단계에서 반영해주세요~ 👍
